src/shortest_path_graph.py

```diff
@@ -0,0 +1,206 @@
+    is complete.
+    g.breadth_first_traversal(start): Perform a full breadth-first traversal
+    of the graph, beginning at start. Return the full visited path when
+    traversal is complete.
+    References:
+    https://www.python.org/doc/essays/graphs/
+    https://medium.freecodecamp.com/a-gentle-introduction-to-data-structures-how-graphs-work-a223d9ef8837#.6xbpr1l6q
+    http://stackoverflow.com/questions/19472530/representing-graphs-data-structure-in-python
+    """
+
+    def __init__(self, data=None):
+        """Initialize a graph instance.
+        Data is the key of the dict and edges are held in a list of names.
+        """
+        self.graph = {}
+        if data:
+            for i in data:
+                self.add_node(i)
+
+    def nodes(self):
+        """Return a list of all nodes in graph."""
+        return [key for key in self.graph.keys()]
+
+    def edges(self):
+        """Return a list of all edges in the graph."""
+        edge_list = []
+        if self.nodes:
+            for node1 in self.graph:
+                for node2 in self.graph[node1]:
+                    edge_list.append((node1, node2[0], node2[1]))
+        return edge_list
+
+    def add_node(self, node):
+        """Add a new node to graph."""
+        self.graph.setdefault(node, [])
+
+    def add_edge(self, node1, node2, weight=1):
+        """Add an edge between node1 and node2."""
+        self.graph.setdefault(node1, [])
+        self.graph.setdefault(node2, [])
+        if node2 not in self.graph[node1]:
+            self.graph[node1].append((node2, weight))
+
+    def del_node(self, node_delete):
+        """Delete the inputted node from the graph."""
+        if node_delete not in self.graph:
+            raise IndexError('The input node is not in the graph')
+        del self.graph[node_delete]
+        for node in self.graph:
+            for edge in self.graph[node]:
+                if edge[0] == node_delete:
+                    self.graph[node].remove(edge)
+
+    def del_edge(self, node1, node2):
+        """Delete the edge connecting node1 to node 2 if it exists."""
+        if node2 not in [edge[0] for edge in self.graph[node1]]:
+            raise IndexError('This edge does not exist.')
+
+        removed_edge = [edge for edge in self.graph[node1] if edge[0] == node2]
+        self.graph[node1].remove(removed_edge[0])
+
+    def has_node(self, node):
+        """Return true if the input node is in the graph, else False."""
+        return node in self.graph
+
+    def neighbours(self, node):
+        """Return the list of nodes connected to the input node."""
+        return [edge[0] for edge in self.graph[node]]
+
+    def adjacent(self, node1, node2):
+        """Return True if there is an edge connecting n1 and n2."""
+        if node1 not in self.graph or node2 not in self.graph:
+            raise KeyError('One or both of these nodes is not in the graph.')
```

Comment:
Though this is a keyerror under the hood, a ValueError would be more appropriate here